### PR TITLE
Block refactor (nio1)

### PR DESCRIPTION
**Pull request by mattdodge**
This is a refactor of the entire socket.io block/client that is in place for n.io 1. It has run at duration for 4-5 days now and seems to be functioning properly.

The main thing that this refactor will fix is the locking around connecting/disconnecting/sending. During periods of downtime or outage the block is smarter about not sending at the wrong time. This in turn causes cleaner reconnections and prevents the block from getting into parallel reconnection attempts.

### CODE REVIEW

**Comment by mattdodge** [comment]
Something's up with the git tree, I have no idea what or how it happened. Only the last 3 commits are in this PR (8535562...0848c2e) but it's showing stuff from the nio2 branch on there too?

Regardless, the result if this PR is merged will definitely be a nio1 block.

File: client/client.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    def handshake_ok(self):
+        """ Called when the initial handshake succeeds.
+
+        This method will start our client thread which will then wait
+        for a successful connection to complete.
+        """
+        self.logger.info("Handshake successful")
+        self._th.start()
+        self._th.join(timeout=1.0)
```
Comment (f1401martin) [inline client/client.py]: Would it make sense for this 1.0 to be configurable?

File: client/receiver.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            self.logger.warning(
+                "Received an improperly formatted message: %s" % m)
+            return
+
+        message_data = str(m)[2:]
+
+        self.logger.debug("Received a message: {}".format(message_data))
+
+        # Handle the different types
+        message_handlers = {
```
Comment (f1401martin) [inline client/receiver.py]: Consider defining this dictionary in __init__

File: client/sender.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import json
+
+
+class PacketSender(object):
+
+    def __init__(self, client, logger):
+        super().__init__()
```
Comment (f1401martin) [inline client/sender.py]: Not needed

File: client/sender.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        self.send_packet(4, '', '', json.dumps(data))
+
+    def send_event(self, event, data):
+        self.send_packet(42, event, data)
+
+    def send_heartbeat(self):
+        self.logger.debug("Sending heartbeat message")
+        self.send_packet(2)
+
+    def send_packet(self, code, path='', data='', id=''):
```
Comment (f1401martin) [inline client/sender.py]: id parameter is not used

File: socketio_block.py [hunk 5]
```diff
@@ -36,63 +28,65 @@ class SocketIO(Retry, Block):
...
-    room = StringProperty(title='SocketIo Room', default="default")
+    room = StringProperty(title='Socket.io Room', default="default")
     content = ExpressionProperty(
-        title='Content', default="{{json.dumps($to_dict(), default=str)}}",
+        title='Content', default="{{ json.dumps($to_dict(), default=str) }}",
         visible=False)
     listen = BoolProperty(title="Listen to SocketIo Room", default=False)
-    socketio_version = SelectProperty(
-        SocketIOVersion, title='Socket.IO Version', default=SocketIOVersion.v1)
+    connect_timeout = TimeDeltaProperty(
```
Comment (f1401martin) [inline socketio_block.py]: Isn't it desired to change this value through the UI, why is visible set to False?

File: socketio_block.py [hunk 6]
```diff
@@ -36,63 +28,65 @@ class SocketIO(Retry, Block):
...
+        self._connection_semaphore = BoundedSemaphore(1)
+        self._socket_url_protocol = "http"
         self._socket_url_base = ""
         self._stopping = False
-        self._reconnecting = False
 
     def configure(self, context):
         super().configure(context)
+        # For nio 2 compatibility
+        self.logger = self._logger
```
Comment (f1401martin) [inline socketio_block.py]: There is usage of both _logger and logger sprinkled throughout

File: client/client.py (hunk 1 above)
Comment (mattdodge) [inline client/client.py]: I'm actually not even sure why this is there haha. I'll see if I can find out.

File: socketio_block.py (hunk 5 above)
Comment (mattdodge) [inline socketio_block.py]: We use visible false for "advanced" properties that the average user won't change. I think this property falls under that. Eventually we'll have some UI features to see the advanced properties but until then we edit them with the json or API.

File: socketio_block.py (hunk 6 above)
Comment (mattdodge) [inline socketio_block.py]: Yeah, I left the `self.logger` references to make the merges into the nio2 branches easier. We say `self.logger` in nio 2 and unfortunately we're maintaining two versions of this block.

**Comment by mattdodge** [comment]
Looking into these changes, thanks Franky

**Pull request merged**